packages/walletconnect/src/index.ts

```diff
@@ -1,121 +1,121 @@
+  private readonly options: Omit<WalletConnectOptions, 'rpcMap' | 'chains'>
+
+  private readonly rpcMap?: Record<number, string | string[]>
+  private readonly chains: number[]
+  private readonly defaultChainId?: number
+
   private readonly timeout: number
 
-  private eagerConnection?: Promise<void>
+  private eagerConnection?: Promise<WalletConnectProvider>
 
-  constructor({ actions, options, defaultChainId, timeout = 5000, onError }: WalletConnectConstructorArgs) {
+  constructor({ actions, options, defaultChainId, timeout = DEFAULT_TIMEOUT, onError }: WalletConnectConstructorArgs) {
     super(actions, onError)
 
-    const { rpc, ...rest } = options
+    const { rpcMap, rpc, chains, ...rest } = options
+
     this.options = rest
-    this.rpc = Object.keys(rpc).reduce<{ [chainId: number]: string[] }>((accumulator, chainId) => {
-      const value = rpc[Number(chainId)]
-      accumulator[Number(chainId)] = Array.isArray(value) ? value : [value]
-      return accumulator
-    }, {})
-    this.defaultChainId = defaultChainId ?? Number(Object.keys(this.rpc)[0])
+    this.chains = chains
+    this.defaultChainId = defaultChainId
+    this.rpcMap = rpcMap || rpc
+
     this.timeout = timeout
   }
 
-  private disconnectListener = (error?: ProviderRpcError): void => {
+  private disconnectListener = (error: ProviderRpcError) => {
     this.actions.resetState()
     if (error) this.onError?.(error)
   }
 
-  private chainChangedListener = (chainId: number | string): void => {
-    this.actions.update({ chainId: parseChainId(chainId) })
+  private chainChangedListener = (chainId: string): void => {
+    this.actions.update({ chainId: Number.parseInt(chainId, 16) })
   }
 
   private accountsChangedListener = (accounts: string[]): void => {
     this.actions.update({ accounts })
   }
 
-  private URIListener = (_: Error | null, payload: { params: string[] }): void => {
-    this.events.emit(URI_AVAILABLE, payload.params[0])
+  private URIListener = (uri: string): void => {
+    this.events.emit(URI_AVAILABLE, uri)
   }
 
-  private async isomorphicInitialize(chainId = this.defaultChainId): Promise<void> {
-    if (this.eagerConnection) return
-
-    // because we can only use 1 url per chainId, we need to decide between multiple, where necessary
-    const rpc = Promise.all(
-      Object.keys(this.rpc).map(
-        async (chainId): Promise<[number, string]> => [
-          Number(chainId),
-          await getBestUrl(this.rpc[Number(chainId)], this.timeout),
-        ]
-      )
-    ).then((results) =>
-      results.reduce<{ [chainId: number]: string }>((accumulator, [chainId, url]) => {
-        accumulator[chainId] = url
-        return accumulator
-      }, {})
-    )
-
-    return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (m) => {
-      this.provider = new m.default({
+  private isomorphicInitialize(
+    desiredChainId: number | undefined = this.defaultChainId
```

Comment:
If a user explicitly passes `undefined`, this will default to `this.defaultChainId`. Is this intended behavior?

Reply:
The only way this gets undefined is when `activate()` is called without a chainId and in this case, we want to trigger `defaultChainId`, which can also be undefined (in this case, we will just default to whatever was first in the array). In that sense, `defaultChainId` is more of a compatibility layer as you can control it via array.